### PR TITLE
storagegroup: Improvements and test coverage

### PR DESCRIPTION
**Pull request by cthulhu-rider**

### CODE REVIEW

File: storagegroup/storagegroup.go [hunk 1]
```diff
@@ -18,13 +18,35 @@ import (
...
+func New(size uint64, cs checksum.Checksum, members []oid.ID) StorageGroup {
+	var res StorageGroup
+	res.SetValidationDataSize(size)
+	res.SetValidationDataHash(cs)
+	res.SetMembers(members)
+	return res
+}
+
+// Unmarshal creates new StorageGroup and makes [Version.Unmarshal].
+func Unmarshal(b []byte) (StorageGroup, error) {
```
Comment (carpawell) [inline storagegroup/storagegroup.go:41]: is it the first time we have just `Unmarshal` func (not a method)? do you plan the other structs to have it too?

Comment (carpawell) [reply storagegroup/storagegroup.go:41]: "Version"?

File: storagegroup/storagegroup.go [hunk 2]
```diff
@@ -234,8 +234,8 @@ func (sg *StorageGroup) SetMembers(members []oid.ID) {
 // Marshal marshals StorageGroup into a protobuf binary form.
 //
 // See also Unmarshal.
-func (sg StorageGroup) Marshal() ([]byte, error) {
-	return (*storagegroup.StorageGroup)(&sg).StableMarshal(nil), nil
+func (sg StorageGroup) Marshal() []byte {
```
Comment (carpawell) [inline storagegroup/storagegroup.go:205]: i would drop all at once. this was is also ok to me

File: storagegroup/storagegroup.go [hunk 3]
```diff
@@ -19,11 +19,13 @@ import (
 // message. See ReadFromMessageV2 / WriteToMessageV2 methods.
 //
 // Instances should be created using one of the constructors.
-//
-// Note that direct typecast is not safe and may result in loss of compatibility:
-//
-//	_ = StorageGroup(storagegroup.StorageGroup) // not recommended
-type StorageGroup storagegroup.StorageGroup
+type StorageGroup struct {
```
Comment (carpawell) [inline storagegroup/storagegroup.go:22]: can you explain the motivation? we have it as a struct but we also have this: https://github.com/nspcc-dev/neofs-sdk-go/pull/598/commits/a07fde16f59c012e8e14a8900f11ba66adf47daf

can you at least explain it in commit? it does not look clear

File: storagegroup/storagegroup.go (hunk 1 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:41]: #483 is for these

File: storagegroup/storagegroup.go (hunk 3 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:22]: those are arrays, this is a struct

> can you at least explain it in commit?

done already

File: storagegroup/storagegroup.go (hunk 2 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:206]: gone after rebase

File: storagegroup/storagegroup.go (hunk 1 above)
Comment (carpawell) [reply storagegroup/storagegroup.go:41]: i mean, will we see such funcs (not methods) for other structs that already have `Unmarshal` method?

File: storagegroup/storagegroup.go (hunk 3 above)
Comment (carpawell) [reply storagegroup/storagegroup.go:22]: > done already

it refs to another commit that says "This privatizes all fields." that is it?

> those are arrays, this is a struct

arrays also have some unique unwanted methods (slicing, element extraction by an index)

File: storagegroup/storagegroup.go (hunk 1 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:41]: we can add them anytime. Im tryin to do this while traversing all packages now. Finally, yes we will

File: storagegroup/test/generate.go [hunk 4]
```diff
@@ -10,7 +10,8 @@ import (
 
 // StorageGroup returns random storagegroup.StorageGroup.
 func StorageGroup() storagegroup.StorageGroup {
-	x := storagegroup.New(322, checksumtest.Checksum(), oidtest.IDs(2))
+	n := 1 + rand.Int()%10
```
Comment (roman-khimov) [inline storagegroup/test/generate.go:13]: `Intn()` would be better even though it's test code and we don't care much about value distribution.

File: storagegroup/storagegroup.go [hunk 5]
```diff
@@ -65,28 +67,26 @@ func (sg *StorageGroup) readFromV2(m storagegroup.StorageGroup, checkFieldPresen
...
-
-			mMembers[member] = struct{}{}
 		}
 	} else if checkFieldPresence {
 		return errors.New("missing members")
 	}
 
-	*sg = StorageGroup(m)
+	sg.sz = m.GetValidationDataSize()
+	//nolint:staticcheck
```
Comment (roman-khimov) [inline storagegroup/storagegroup.go:88]: Why is it complaining?

File: crypto/test/tests.go [hunk 6]
```diff
@@ -52,7 +52,7 @@ func FailSigner(s neofscrypto.Signer) neofscrypto.Signer {
 
 // Signature returns random neofscrypto.Signature.
 func Signature() neofscrypto.Signature {
-	sig := make([]byte, rand.Int()%128)
+	sig := make([]byte, 1+rand.Int()%128)
```
Comment (roman-khimov) [inline crypto/test/tests.go:55]: Suggest `Intn()` here as well.

**Review by roman-khimov — APPROVED** [review]

File: storagegroup/storagegroup.go (hunk 3 above)
Comment (carpawell) [reply storagegroup/storagegroup.go:22]: > not sure i got ur point right

you said that the difference b/w `StorageGroup` and `user.ID` is structs vs arrays. but arrays can also be misused (they have indexes, etc). i don't mind either structs or just new types, just want to understand why sometimes it is done one way, but at a different time it is done another way (and commit message can help)

**Review by carpawell — APPROVED** [review]

File: storagegroup/storagegroup.go (hunk 5 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:88]:
```
storagegroup/storagegroup.go:88:11  staticcheck  SA1019: m.GetExpirationEpoch is deprecated: Do not use.
```

File: storagegroup/test/generate.go (hunk 4 above)
Comment (cthulhu-rider) [reply storagegroup/test/generate.go]: done

File: crypto/test/tests.go (hunk 6 above)
Comment (cthulhu-rider) [reply crypto/test/tests.go]: done

File: storagegroup/storagegroup.go (hunk 3 above)
Comment (cthulhu-rider) [reply storagegroup/storagegroup.go:22]: > commit message can help

idk still what to write in it

**Pull request merged**